tests/test_bootloader.py

```diff
@@ -0,0 +1,634 @@
+
+  Parameters
+  ==========
+  port1 : string
+    Filepath to the COM port of the Piksi Under Test.
+  port2 : string
+    Filepath to the COM port of a Piksi connected via UART to the
+    Piksi Under Test.
+  stm_fw : intelhex.IntelHex
+    STM firmware to use in tests.
+  nap_fw : intelhex.IntelHex
+    NAP firmware to use in tests.
+  verbose : bool
+    Print status output.
+  n_queue : int
+    Number of flash operations to queue.
+  """
+  if verbose: print "--- Setting up device for tests ---"
+
+  with serial_link.get_driver(use_ftdi=False, port=port1) as driver:
+    with Handler(driver.read, driver.write) as handler:
+      setup_piksi(handler, stm_fw, nap_fw, verbose)
+
+  if verbose: print ""
+
+  # Hack: Wait a bit for 'Piksi Disconnected'
+  # print from sbp.client.handler.Handler
+  time.sleep(0.1)
+
+class TestBootloader(unittest.TestCase):
+
+  def __init__(self, testname, port1, port2, stm_fw, nap_fw, verbose, \
+               n_queue, skip_single, skip_double):
+    """
+    Piksi bootloader tests. Tests assume that Piksies have a valid bootloader
+    and STM / NAP firmware, such that it sends bootloader handshake messages and
+    will enter bootloader mode upon receiving a bootloader handshake message,
+    sends heartbeat messages, and can be reset through receiving a reset message.
+    All tests should leave device in this state upon returning.
+
+    Parameters
+    ==========
+    testname : string
+      Test to instantiate.
+    port1 : string
+      Filepath to the COM port of the Piksi Under Test.
+    port2 : string
+      Filepath to the COM port of a Piksi connected via UART to the
+      Piksi Under Test.
+    stm_fw : intelhex.IntelHex
+      STM firmware to use in tests.
+    nap_fw : intelhex.IntelHex
+      NAP firmware to use in tests.
+    verbose : bool
+      Print status output.
+    n_queue : int
+      Number of flash operations to queue.
+    skip_single : bool
+      Skip single Piksi tests.
+    skip_double : bool
+      Skip double Piksi tests.
+    """
+    super(TestBootloader, self).__init__(testname)
+    self.port1 = port1
+    self.port2 = port2
+    self.stm_fw = stm_fw
+    self.nap_fw = nap_fw
+    self.verbose = verbose
+    self.n_queue = n_queue
+    self.skip_single = skip_single
+    self.skip_double = skip_double
+
+  def tearDown(self):
+    """ Clean up after running each test. """
+    # Hack: Wait a bit for 'Piksi Disconnected'
+    # print from sbp.client.handler.Handler
+    time.sleep(0.1)
+
+  def _piksi_settings_cb(self):
+    """ Callback to set flag indicating Piksi settings have been received. """
+    self.settings_received = True
+
+  def get_piksi_settings(self, handler):
+    """
+    Get settings from Piksi.
+
+    Parameters
+    ==========
+    handler : sbp.client.handler.Handler
+      Handler to send messages to Piksi over and register callbacks with.
+    """
+
+    self.settings_received = False
+
+    with SettingsView(handler,
```

Comment:
Maybe not in this PR, but we should move this on to `piksi_tools.Diagnostics` and a `Diagnostics(link).diagnostics` call.